proxy/proxy.go

```diff
@@ -0,0 +1,174 @@
+package proxy
+
+import (
+	"fmt"
+	"io"
+	"log"
+	"net"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/gorilla/mux"
+)
+
+const (
+	watchdogPort       = "8080"
+	defaultContentType = "text/plain"
+)
+
+// BaseURLResolver URL resolver for proxy requests
+//
+// The FaaS provider implementation is responsible for providing the resolver function implementation.
+// BaseURLResolver.Resolve will receive the function name and should return the base Address of the
+// function service.
+type BaseURLResolver interface {
+	Resolve(functionName string) (string, error)
+}
+
+// NewHandlerFunc creates a standard http HandlerFunc to proxy function requests to the function.
+func NewHandlerFunc(timeout time.Duration, resolver BaseURLResolver) http.HandlerFunc {
+	proxyClient := http.Client{
+		Transport: &http.Transport{
+			Proxy: http.ProxyFromEnvironment,
+			DialContext: (&net.Dialer{
+				Timeout:   timeout,
+				KeepAlive: 1 * time.Second,
```

Comment:
@LucasRoesler i'm curious I've seen these adjustments to the default transport in `faas-cli` as well

- DialContex.KeepAlive
- IdleConnTimeout
- ExpectContinueTimeout

The differ substantially from `DefaultTransport` is there a way to infer to the reader of the code why we are doing this?

Reply:
I could add some comments. The core reason, that I understand, for these settings is that some are just not set by the default transport and client

Reply:
A common issue with Golang's HTTP library was inappropriate defaults which lead to a "DoS" attack or infinite timeouts when doing performance testing/benchmarking. We tuned these values but the Golang team has also improved upon their defaults since we made these changes originally. cc @stefanprodan

Reply:
You should also see some context in the git history in the faas repo where this code was taken from.

Reply:
Please take a look at the newly added comment and let me know if more context would be useful